internal/pkg/v1/action/catalog_create.go

```diff
@@ -0,0 +1,81 @@
+package action
+
+import (
+	"context"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	olmv1catalogd "github.com/operator-framework/catalogd/api/v1"
+
+	"github.com/operator-framework/kubectl-operator/pkg/action"
+)
+
+type CatalogCreate struct {
+	config         *action.Configuration
+	CatalogName    string
+	ImageSourceRef string
+
+	Priority            int32
+	PollIntervalMinutes int
+	Labels              map[string]string
+	Available           bool
+	CleanupTimeout      time.Duration
+
+	Logf func(string, ...interface{})
+}
+
+func NewCatalogCreate(cfg *action.Configuration) *CatalogCreate {
+	return &CatalogCreate{
+		config: cfg,
+		Logf:   func(string, ...interface{}) {},
+	}
+}
+
+func (i *CatalogCreate) Run(ctx context.Context) error {
+	catalog := i.buildCatalog()
+	if err := i.config.Client.Create(ctx, &catalog); err != nil {
+		return err
+	}
+
+	var err error
+	if i.Available {
+		err = waitUntilCatalogStatusCondition(ctx, i.config.Client, &catalog, olmv1catalogd.TypeServing, metav1.ConditionTrue)
+	} else {
+		err = waitUntilCatalogStatusCondition(ctx, i.config.Client, &catalog, olmv1catalogd.TypeServing, metav1.ConditionFalse)
+	}
+
+	if err != nil {
+		if cleanupErr := deleteWithTimeout(i.config.Client, &catalog, i.CleanupTimeout); cleanupErr != nil {
```

Comment:
With this cleanup, I'm following what olmv0 commands are doing.

Treating create operation as atomic and deleting the failed catalog behind the scenes probably makes the most sense, though it might not be ideal from debugging POV, because the actual reason for failed catalog would be hidden within status conditions of the CR we will be cleaning up here. We are also introducing additional waiting time until the command returns.
User can still run this again and track the status, so not a big deal, but similarly to status waiting above - curious about what people think.

Reply:
> though it might not be ideal from debugging POV, because the actual reason for failed catalog would be hidden within status conditions of the CR we will be cleaning up here.

I think this can _mostly_ be mitigated by having good reporting logic that happens prior to the cleanup, so we could tell a user what went wrong. In the case where our logic (or the CE status) just doesn't provide enough context, there's always logs from the operator-controller pod (assuming the user has access and knows where/how to look for them).